### PR TITLE
support multiple listeners with multiple endpoints

### LINKED ISSUE #574: [FEATURE] Support multiple listeners of one protocol
**Is your feature request related to a problem? Please describe.**

Currently for KoP we can configure `kafkaListeners` as what interfaces KoP binds to and `kafkaAdvertisedListeners` as how clients connect. However, currently KoP can only bind to one interface for each protocol.

For example, if KoP was configured with

```properties
kafkaListeners=PLAINTEXT://127.0.0.1:9092,SSL://127.0.0.1:9093
```

it would work well.

But if KoP was configured with

```properties
kafkaListeners=PLAINTEXT://127.0.0.1:9092,SSL://127.0.0.1:9093,SSL://127.0.0.1:9094
```

KoP would fail to start with following logs.

```
15:56:29.408 [main] ERROR io.streamnative.pulsar.handlers.kop.KafkaProtocolHandler - KafkaProtocolHandler newChannelInitializers failed with 
java.lang.IllegalStateException: PLAINTEXT://127.0.0.1:9092,SSL://127.0.0.1:9093,SSL://127.0.0.1:9094 has multiple listeners whose protocol is SSL
	at io.streamnative.pulsar.handlers.kop.EndPoint.parseListeners(EndPoint.java:77) ~[pulsar-protocol-handler-kafka-2.8.0-rc-202106091215.nar-unpacked/:?]
	at io.streamnative.pulsar.handlers.kop.KafkaProtocolHandler.newChannelInitializers(KafkaProtocolHandler.java:313) [pulsar-protocol-handler-kafka-2.8.0-rc-202106091215.nar-unpacked/:?]
	at org.apache.pulsar.broker.protocol.ProtocolHandlerWithClassLoader.newChannelInitializers(ProtocolHandlerWithClassLoader.java:71) [io.streamnative-pulsar-broker-2.8.0-rc-202106071430.jar:2.8.0-rc-202106071430]
	at org.apache.pulsar.broker.protocol.ProtocolHandlers.newChannelInitializers(ProtocolHandlers.java:124) [io.streamnative-pulsar-broker-2.8.0-rc-202106071430.jar:2.8.0-rc-202106071430]
	at org.apache.pulsar.broker.PulsarService.start(PulsarService.java:787) [io.streamnative-pulsar-broker-2.8.0-rc-202106071430.jar:2.8.0-rc-202106071430]
	at org.apache.pulsar.PulsarStandalone.start(PulsarStandalone.java:296) [io.streamnative-pulsar-broker-2.8.0-rc-202106071430.jar:2.8.0-rc-202106071430]
	at org.apache.pulsar.PulsarStandaloneStarter.main(PulsarStandaloneStarter.java:121) [io.streamnative-pulsar-broker-2.8.0-rc-202106071430.jar:2.8.0-rc-202106071430]
```

**Describe the solution you'd like**
Kafka (as of [KIP-103](https://cwiki.apache.org/confluence/display/KAFKA/KIP-103%3A+Separation+of+Internal+and+External+traffic#KIP103:SeparationofInternalandExternaltraffic-Protocol)) has the [listener.security.protocol.map](https://kafka.apache.org/documentation/#brokerconfigs_listener.security.protocol.map) config to ~map your own protocol~ define a set of listeners, each with a name and a security protocol (`PLAINTEXT`, `SSL`), so that a protocol like SSL could bind multiple ports or IPs.

We can also see more details in [this blog](https://www.confluent.io/blog/kafka-client-cannot-connect-to-broker-on-aws-on-docker-etc/#adding-new-listener). Here's an example to advertise multiple ports of PLAINTEXT protocol:

```properties
advertised.listeners=PLAINTEXT://broker:9092,CONNECTIONS_FROM_HOST://localhost:19092
listener.security.protocol.map=PLAINTEXT:PLAINTEXT,CONNECTIONS_FROM_HOST:PLAINTEXT
```

**Describe alternatives you've considered**

**Additional context**

### LINKED ISSUE #669: [FEATURE] Support multiple listeners by using advertisedListeners
**Is your feature request related to a problem? Please describe.**
Here provide a design for #574, could also be a solution for #633 

Suppose we deploy kop with pulsar on a host which have two NIC, such as a cloud ECS. 192.168.1.10 is the address for private network inside the cloud and 172.16.1.10 is the address for public network.
We start pulsar endpoints at 6650, 6651 on both NIC for pulsar, and start endpoints at 9092, 9093, 19002, 19003 for kafka.

We would config like this for pulsar:
> internalListenerName=pulsar_private
> advertisedListeners=pulsar_private:pulsar://192.168.1.10:6650,pulsar_private_ssl:pulsar+ssl://192.168.1.10:6651,pulsar_public:pulsar://172.16.1.10:6650,pulsar_pulic_ssl:pulsar+ssl://172.16.1.10:6651

We would config like this for kafka:

> listeners=PRIVATE://0.0.0.0:9092,PRIVATE_SSL://0.0.0.0:9093,PUBLIC://0.0.0.0:19002,PUBLIC_SSL:0.0.0.0:19003
> advertised.listeners=PRIVATE://192.168.1.10:9092,PRIVATE_SSL://192.168.1.10:9093,PUBLIC://172.16.1.10:19002,PUBLIC_SSL:172.16.1.10:19003
> listener,security.protocal.map=PRIVATE:PLAINTEXT,PRIVATE_SSL:SSL,PUBLIC:PLAINTEXT,PUBLIC_SSL:SSL

Kop current use `kafkaAdvertisedListeners` to config advertised address for kafka, but only support one address of one protocol.

**Describe alternatives you've considered**
Deprecated the setting `kafkaAdvertisedListeners`.
remove the function `KafkaRequestHandler.getProtocolDataToAdvertise()`
and config kop like this:

> kafkaListeners=PRIVATE://0.0.0.0:9092,PRIVATE_SSL://0.0.0.0:9093,PUBLIC://0.0.0.0:19002,PUBLIC_SSL:0.0.0.0:19003
> kafkaProtocalMap=PRIVATE:PLAINTEXT,PRIVATE_SSL:SSL,PUBLIC:PLAINTEXT,PUBLIC_SSL:SSL
> advertisedListeners={pulsar's listeners},PRIVATE:pulsar//192.168.1.10:9092,PRIVATE_SSL:pulsar//192.168.1.10:9093,PUBLIC:pulsar//172.16.1.10:19002,PUBLIC_SSL:pulsar:172.16.1.10:19003
> 

We should create four endpoints in this case mapped port with listenerName: `PRIVATE:9092,PRIVATE_SSL:9093,PUBLIC:19002,PUBLIC_SSL:19003`

When we receive a request from client, we get the listener name mapped with the endpoint which client connected to, use the listener name to do topic lookup and get the advertised address configed in  `advertisedListeners`, get the protocal mapped with the listener name configed in `kafkaProtocalMap`, then we could generate the lookup response to client.

### PR DESCRIPTION
**Pull request by wangjialing218**
Fixes #669
Fixes #574

To support multiple listeners with multiple protocols, we need to set follow configration, and deprecated `kafkaAdvertisedListeners` and `kafkaListenerName`.

> kafkaListeners=internal://0.0.0.0:9092,internal_ssl://0.0.0.0:9093,external://0.0.0.0:19002,external_ssl:0.0.0.0:19003
> kafkaProtocalMap=internal:PLAINTEXT,internal_ssl:SSL,external:PLAINTEXT,external_ssl:SSL
> advertisedListeners={pulsar's listeners},internal:pulsar//192.168.1.10:9092,internal_ssl:pulsar//192.168.1.10:9093,external:pulsar//172.16.1.10:19002,external_ssl:pulsar://172.16.1.10:19003

1. Define the listenerNames and port of each listenerName in `kafkaListeners`
2. Define the listenerNames and protocol of each listenerName in `kafkaProtocalMap`
3. Add listenerNames and advertised address of each listenerName into `advertisedListeners`

Kafka client should connect to the port of the listenerName according to KIP-103.

### CODE REVIEW

**Comment by wangjialing218** [comment]
@BewareMyPower please have a look. 
I'm not sure if `KopBrokerLookupManager.findBroker` need change, it seems only used in `TransactionMarkerChannelManager`. If the lookup result here is only used in kop (not for kafka client), we could just use buildin client to do lookup.

**Comment by BewareMyPower** [comment]
Could you solve the conflicts?

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/EndPoint.java [hunk 1]
```diff
@@ -31,26 +31,45 @@
...
-    public EndPoint(final String listener) {
+    public EndPoint(final String listener, final String kafkaProtocolMap) {
         this.originalListener = listener;
         final String errorMessage = "listener '" + listener + "' is invalid";
         final Matcher matcher = PATTERN.matcher(listener);
         checkState(matcher.find(), errorMessage);
         checkState(matcher.groupCount() == 3, errorMessage);
 
-        this.securityProtocol = SecurityProtocol.forName(matcher.group(1));
+        Map<String, SecurityProtocol> protocolMap = parseProtocolMap(kafkaProtocolMap);
```
Comment (BewareMyPower) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/EndPoint.java:59]: We can pass a `Map<String, SecurityProtocol>` to the constructor's parameter to avoid calling `parseProtocolMap` every time a `EndPoint` is constructed.

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaListenerNameTest.java [hunk 2]
```diff
@@ -61,4 +65,47 @@ public void testListenerName() throws Exception {
...
+        props.setProperty(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
+
+        final KafkaProducer<String, String> producer = new KafkaProducer<>(props);
+        producer.send(new ProducerRecord<>("my-topic", "my-message"), (metadata, exception) -> {
+            if (exception == null) {
+                log.info("Send to {}", metadata);
+            } else {
+                log.error("Send failed: {}", exception.getMessage());
+            }
+        });
```
Comment (BewareMyPower) [inline tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaListenerNameTest.java:108]: We should use a synchronous send here to propagate the exception if send failed. Here it call `send()`, which is asynchronous, and then close the producer immediately, it cannot guarantee the send is completed.

**Review by BewareMyPower — CHANGES_REQUESTED** [review]

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaProtocolHandler.java [hunk 3]
```diff
@@ -321,7 +321,7 @@ public void start(BrokerService service) {
             KopVersion.getBuildHost(),
             KopVersion.getBuildTime());
 
-        // Currently each time getMetadataCache() is called, a new MetadataCache<T> instance will be created, even for
+        // Currently each  time getMetadataCache() is called, a new MetadataCache<T> instance will be created, even for
```
Comment (BewareMyPower) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaProtocolHandler.java:324]: Revert this change (redundant space)

**Comment by BewareMyPower** [comment]
> I'm not sure if `KopBrokerLookupManager.findBroker` need change, it seems only used in `TransactionMarkerChannelManager`. If the lookup result here is only used in kop (not for kafka client), we could just use buildin client to do lookup.

Regarding to this question, I think we can ignore the transaction part in this PR.

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaListenerNameTest.java (hunk 2 above)
Comment (BewareMyPower) [reply tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaListenerNameTest.java:108]: A simple fix is

```java
producer.send(new ProducerRecord<>("my-topic", "my-message")).get();
```

if you don't want to print the logs.

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/EndPoint.java (hunk 1 above)
Comment (wangjialing218) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/EndPoint.java]: done

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaListenerNameTest.java (hunk 2 above)
Comment (wangjialing218) [reply tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaListenerNameTest.java]: done

**Review by BewareMyPower — APPROVED** [review]
